middle_end/flambda/types/env/aliases.ml

```diff
@@ -126,161 +128,288 @@ end = struct
       Name_mode.Map.merge (fun _order elts1 elts2 ->
           match elts1, elts2 with
           | None, None | Some _, None | None, Some _ -> None
-          | Some elts1, Some elts2 -> Some (Simple.Set.inter elts1 elts2))
+          | Some elts1, Some elts2 -> Some (Name.Set.inter elts1 elts2))
         t1.aliases t2.aliases
     in
     let t =
       { aliases;
-        all = Simple.Set.inter t1.all t2.all;
+        all = Name.Set.inter t1.all t2.all;
       }
     in
     invariant t;
     t
 
-  let rename rename_simple { aliases; all } =
+  let rename rename_name { aliases; all } =
     let aliases =
-      Name_mode.Map.map (fun elts -> Simple.Set.map rename_simple elts)
+      Name_mode.Map.map (fun elts -> Name.Set.map rename_name elts)
         aliases
     in
-    let all = Simple.Set.map rename_simple all in
+    let all = Name.Set.map rename_name all in
     { aliases; all }
 
   let merge t1 t2 =
     let aliases =
       Name_mode.Map.union (fun _mode set1 set2 ->
-          Some (Simple.Set.union set1 set2))
+          Some (Name.Set.union set1 set2))
         t1.aliases
         t2.aliases
     in
-    let all = Simple.Set.union t1.all t2.all in
+    let all = Name.Set.union t1.all t2.all in
     { aliases; all; }
 end
 
+module Alias_set = struct
+  type t = {
+    const : Const.t option;
+    names : Name.Set.t;
+  }
+
+  let empty = { const = None; names = Name.Set.empty; }
+
+  let singleton simple =
+    Simple.pattern_match simple
+      ~const:(fun const ->
+        { const = Some const; names = Name.Set.empty; })
+      ~name:(fun name ->
+        { const = None; names = Name.Set.singleton name })
+
+  let get_singleton { const; names; } =
+    match const with
+    | Some const ->
+      if Name.Set.is_empty names then Some (Simple.const const) else None
+    | None ->
+      Name.Set.get_singleton names
+      |> Option.map Simple.name
+
+  let print ppf { const; names; } =
+    let none ppf () =
+      Format.fprintf ppf "@<0>%s()" (Flambda_colours.elide ())
+    in
+    Format.fprintf ppf
+      "@[<hov 1>(\
+           @[<hov 1>(const@ %a)@]@ \
+           @[<hov 1>(names@ %a)@]@ \
+       @]"
+       (Format.pp_print_option Const.print ~none) const
+       Name.Set.print names
+
+  let inter
+        { const = const1; names = names1; }
+        { const = const2; names = names2; } =
+    let const =
+      match const1, const2 with
+      | Some const1, Some const2 when Const.equal const1 const2 -> Some const1
```

Comment:
If I understood correctly [inter] will be called on two [Alias_set] mapping to the same alias. In that case it seems to me that having two different const is an error. We should perhaps think of raising if that's the case.

Reply:
Actually, no; this is called from [Type_descr.join], so there's no reason to assume they're equal.